Feature/101772588 bobuhl make go-signalfx more resilient

robert-uhl opened the pull request.

Commit:
Support _metric_ prefix, not _dimension_ prefix

Commit:
Add example of metric prefix use

michaeltrobinson committed:
Modified bucket logic to reset after each report
Also made it configurable so you can disable the reporting of specific metrics in it

michaeltrobinson committed:
Fixed bucket nil map

michaeltrobinson committed:
Simplified bucket metric disabling logic

michaeltrobinson committed:
Reordered datapoint creation in buckets so count isn't reset before min/max

michaeltrobinson committed:
Unexported the disabled metrics map

michaeltrobinson committed:
Added test for disabling metrics in bucket

michaeltrobinson committed:
Added initial sizing to datapoints map to avoid unnecessary allocations

Commit:
Simplify counter handling

Committing now in order to merge Michael's improved bucket handling in.
More work to come.

Commit:
Merge remote-tracking branch 'origin/refactor_buckets' into feature/101772588_bobuhl_make_go_signalfx_more_resilient

Commit:
Initial work on replacing DataPoint abstraction

First few tests now pass.  Much more to go.

Commit:
Massive changes, adding metrics and datapoints

This is a partially-structured changed, adding a plethora of new types
named similarly to existing types.  This is _not_ a complete change!

All tests in reporter_test.go pass, and all that _have_ to be updated
_have_ been.  There's more to do yet, though.

Commit:
Rename SetterCounter to WrappedCounter

Commit:
Add WrappedGauge and WrappedCumulativeCounter

And partially-update the example.  It'll need a further update once the
API is firmed up.

Commit:
Radically revamp architecture

This update firmly distinguishes between metrics and data points.

Still outstanding:

- what to do about floating-point values (this should be very easy to
add back in)

- remove old DataPoint structure and rename dataPoint to DataPoint

Commit:
Replace old DataPoint with new DataPoint

Of note, this removes support for floating-point data point values.
We've determined that we do not need them in our use case; they could be
conveniently added back by creating a DataPoint interface based on a
protoDataPoint method which returns an sfx.DataPoint with the
appropriate members set.

Also of note, this version ensures that Counters, CumulativeCounters &
Gauges are goroutine-safe by hiding their metric and dimension values
from client code.  Versions of the old SetDimension/RemoveDimension
&c. were _not_ written, in the belief that dynamic adjustment of
dimensions is not something likely to be necessary.  Client code should
untrack metrics and track new ones if changes to dimensions are
required.

michaeltrobinson committed:
Changed constants in bucket to be actual consts

Commit:
Document all new methods, functions and types

Includes usage notes for client code.

Commit:
Fix bug in Bucket data point timestamps

Added test to prevent regression, of course.

Commit:
Merge branch 'feature/101772588_bobuhl_make_go_signalfx_more_resilient' of github.com:zvelo/go-signalfx into feature/101772588_bobuhl_make_go_signalfx_more_resilient

michaeltrobinson committed:
Added rollup dimensions back to Buckets

Commit:
Improve test coverage

And fix a few bugs that improved tests turned up.

Commit:
Merge branch 'feature/101772588_bobuhl_make_go_signalfx_more_resilient' of github.com:zvelo/go-signalfx into feature/101772588_bobuhl_make_go_signalfx_more_resilient

Commit:
Update README for new version

Commit:
Fix bug, add test to prevent detect it

rajnmithun commented on counter.go:
```diff
@@ -0,0 +1,127 @@
+// values.
+func (c *Counter) PostReportHook(v int64) {
+	if v < 0 {
+		panic("negative counter should be impossible")
```
@zvbuhl Is it good for a package to panic which would crash the program? or instead return error for the caller. I was thinking common go idiom is to return errors.

robert-uhl replied:
Yeah, I agree that normally in Go one would normally return an error and let the caller figure out how to deal with it. Panics should be reserved to dealing with unrecoverable states.

I think that this qualifies as an unrecoverable state: the argument to a HookedMetric's PostReportHook method is the value return in that HookedMetric's DataPoint method; if that's not a valid value for that HookedMetric, something has gone seriously badly awry. I believe that if something's gone that wrong, then the only thing to do is die and let the orchestration system spin up a new process and alert ops that something weird happened.

There's also the issue that it's unclear how Reporter.Report (which is the only caller of PostReportHook) should recover: should it delete the HookedMetric; should it try to create a new one; should it call some caller-provided recovery function to try to Do The Right Thing in an application-specific manner (and really, 'some caller-provided recovery function' is what panic & recover are all about)?

And then there's the issue that the usual use case is not for clients to directly call Reporter.Report, but rather Reporter.RunInBackground: what should it do if a particular HookedMetric is corrupted? Should it try to handle the issue? Should it pass the value on some sort of error channel back to code running in the main module? Should it call some error-recovery function defined by the main module?

I _think_ that these questions are basically impossible to answer, and thus that this is truly a fatal error, and thus—being a fatal error—it is appropriate to die. But I could be wrong☺

joshuarubin replied:
Good catch @rajnmithun. My rule of thumb is to panic for things that are developer error, e.g. if you create a template that doesn't compile, it should panic at the start of execution (http://golang.org/pkg/text/template/#Must). However, it is probably not a good idea to panic for weird runtime situations, because what we think is impossible as developers has a way of becoming possible in production.

Sure it is possible to recover from panic, but I doubt that people using this package would know that they need to prepare for possible panics. As a result, this could crash production systems which is definitely a blocking issue.

robert-uhl replied:
It is definitely ugly, but what's the alternative? That's an honest question. I only see two:
- Just do nothing if Reporter.Report tells a counter that it just reported a negative value. This has the virtue of not crashing a production system. But if this occurs, that production system's memory is hosed, in which case it's untrustworthy and should crash (and be replaced…) anyway.
- Implement a very complex error-recovery process where the definition of 'recovery' is punted in some way all the way up to the code which calls Reporter.RunInBackground. This has the virtue of enabling some sort of recovery (maybe the main program could choose to remove or recreate the counter or cumulative counter?), although since it could only be the result of memory corruption, the 'recovered' process is likely to crash again soon.

I think that the key point with this code is that it is detecting a memory corruption error; it's the equivalent of `if (*a != *a) …` in unthreaded C. There is absolutely no way that `Counter.DataPoint`, `WrappedCounter.DataPoint`, `CumulativeCounter.DataPoint` or `WrappedCumulativeCounter.DataPoint` will ever produce a negative `DataPoint.Value`, and thus if `Counter.PostReportHook`, `WrappedCounter.PostReportHook`, `CumulativeCounter.PostReportHook` or `WrappedCumulativeCounter.PostReportHook` ever receives a value less than zero, then it's an exceptionally grave runtime error.

If one traces the value, it is produced by one of the `DataPoint` methods of a struct, recorded in a `struct{HookedMetric,int64}` by `Reporter.Report`, then handed back to the original type's `PostReportHook`: it's internal assertion checking.

I guess another way of looking at it is that this is the moral equivalent of a slice bounds violation or nil pointer access: when it happens, the system just can't be trusted to recover cleanly.

I guess I could just remove the check, on the grounds that `if 2 == 1` will never happen and the cost of maintaining the check is higher than it could be worth…

rajnmithun replied:
I was just reading into gRPC server code and found something that might be similar use case for the counter issue     https://github.com/grpc/grpc-go/blob/master/server.go#L239-L247
where they exit the program instead of panic, may be that seems a better way of handling where the program exits, this way it seems that there was nothing to recover from by the user since the system is in corrupted state so it logged and stopped.

Commit:
Merge branch 'develop' into feature/101772588_bobuhl_make_go_signalfx_more_resilient

Commit:
Document panics

Violated function preconditions can cause panics in certain cases.  This
commit documents these cases in detail.

Commit:
Return errors instead of panicking

Where possible.  This is a minor API change, but it only adds a return
value to a function which previously had none.

Commit:
Substitute gogo/protobuf with golang/protobuf

Commit:
Merge branch 'feature/101772588_bobuhl_make_go_signalfx_more_resilient' of github.com:zvelo/go-signalfx into feature/101772588_bobuhl_make_go_signalfx_more_resilient

Conflicts:
	sfxproto/signalfx_test.go
